src/Compiler/Microsoft.AspNetCore.Razor.Language/src/MetadataCollection.cs

```diff
@@ -0,0 +1,721 @@
+                var index = ~keys.BinarySearch(key);
+
+                keys.Insert(index, key);
+                values.Insert(index, value);
+            }
+
+            _keys = keys.ToArrayOrEmpty();
+            _values = values.ToArrayOrEmpty();
+            _count = count;
+        }
+
+        public FourOrMoreItems(KeyValuePair<string, string>[] pairs)
+        {
+            if (pairs is null)
+            {
+                throw new ArgumentNullException(nameof(pairs));
+            }
+
+            using var _1 = ListPool<string>.GetPooledObject(out var keys);
+            using var _2 = ListPool<string>.GetPooledObject(out var values);
+
+            var count = pairs.Length;
+            keys.SetCapacityIfLarger(count);
+            values.SetCapacityIfLarger(count);
+
+            foreach (var (key, value) in pairs)
+            {
+                var index = keys.BinarySearch(key);
+
+                if (index >= 0)
+                {
+                    throw new ArgumentException(
+                        Resources.FormatAn_item_with_the_same_key_has_already_been_added_Key_0(key), nameof(pairs));
+                }
+
+                index = ~index;
+
+                keys.Insert(index, key);
+                values.Insert(index, value);
+            }
+
+            _keys = keys.ToArrayOrEmpty();
+            _values = values.ToArrayOrEmpty();
+            _count = count;
+        }
+
+        public override string this[string key]
+        {
+            get
+            {
+                var index = Array.BinarySearch(_keys, key);
+
+                return index >= 0
+                    ? _values[index]
+                    : throw new KeyNotFoundException(Resources.FormatThe_given_key_0_was_not_present(key));
+            }
+        }
+
+        public override IEnumerable<string> Keys => _keys;
+        public override IEnumerable<string> Values => _values;
+
+        public override int Count => _count;
+
+        public override bool ContainsKey(string key)
+        {
+            var index = Array.BinarySearch(_keys, key);
+
+            return index >= 0;
+        }
+
+        public override IEnumerator<KeyValuePair<string, string>> GetEnumerator()
+        {
+            _pairs ??= CreatePairs(_keys, _values);
+
+            return _pairs.GetEnumerator();
+
+            static IEnumerable<KeyValuePair<string, string>> CreatePairs(string[] keys, string[] values)
+            {
+                for (var i = 0; i < keys.Length; i++)
+                {
+                    yield return new KeyValuePair<string, string>(keys[i], values[i]);
+                }
+            }
+        }
+
+        public override bool TryGetValue(string key, out string value)
+        {
+            var index = Array.BinarySearch(_keys, key);
+
+            if (index >= 0)
+            {
+                value = _values[index];
+                return true;
+            }
+
+            value = null!;
```

Comment:
You could use `[MaybeNullWhen(returnValue: false)]` on the `string value` parameter and avoid these `!`s.

Reply:
I initially had that but it introduced a nullability warning because this is implementing `IReadOnlyDictionary<string, string>.TryGetValue(...)`, which is not annotated on `netstandard2.0`. 

![image](https://user-images.githubusercontent.com/116161/228887339-d7aea56e-2617-49ac-ad2c-ce4f1e023f9b.png)

Since consumers of the tag helper API only consume this method through `IReadOnlyDictionary<string, string>`, I didn't bother tracking this down further to see if I could make it sort of work.